frontend/apps/marketing/cicd/1-setup/setup.rb

```diff
@@ -130,100 +163,182 @@ def process_template(template_file, output_file, binding_object)
     exit 1
   end
 end
 
-def deploy_stack(stack_name:, template_file:, parameters: {}, region:, role_arn: nil, tags: {}, capabilities: [])
-  temp_dir = File.join(Dir.pwd, 'tmp')
-  FileUtils.mkdir_p(temp_dir)
-  parameter_path = nil  # Initialize to nil
-
-  # Build the AWS CLI command
-  command_parts = [
-    "aws cloudformation deploy",
-    "--stack-name #{stack_name}",
-    "--template-file #{template_file}",
-    "--region #{region}"
-  ]
-
-  # Add capabilities if any are specified
-  unless capabilities.empty?
-    command_parts << "--capabilities #{capabilities.join(' ')}"
-  end
+def stack_exists?(stack_name, cloudformation_client)
+  response = cloudformation_client.describe_stacks(stack_name: stack_name)
+  stack = response.stacks.first
+  # Check if stack exists and is not in a deleted state
+  return stack && !['DELETE_COMPLETE'].include?(stack.stack_status)
+rescue Aws::CloudFormation::Errors::ValidationError => exception
+  # Stack doesn't exist
+  return false if exception.message.include?('does not exist')
+  raise exception
+rescue Aws::CloudFormation::Errors::ServiceError => exception
+  puts "Error checking stack existence: #{exception.message}"
+  return false
+end
 
-  command_parts << "--role-arn #{role_arn}" if role_arn
+def deploy_stack(stack_name:, template_file:, parameters: {}, region:, cloudformation_role_arn: nil, tags: {}, capabilities: [], template_s3_bucket:, clients:)
+  s3_client = clients[:s3]
+  cf_client = clients[:cloudformation]
 
-  # Add parameters if any
-  unless parameters.empty?
-    param_file = "parameters_#{stack_name}_#{Time.now.to_i}.json"
-    parameter_path = File.join(temp_dir, param_file)
-    parameter_content = parameters.map {|k, v| {"ParameterKey" => k, "ParameterValue" => v.to_s}}
-    File.write(parameter_path, JSON.pretty_generate(parameter_content))
-    command_parts << "--parameter-overrides file://#{parameter_path}"
-  end
+  # Always upload template to S3
+  template_url = upload_template_to_s3(template_file, template_s3_bucket, s3_client)
+
+  # Check if stack exists to determine create vs update
+  stack_exists = stack_exists?(stack_name, cf_client)
+  operation = stack_exists ? :update : :create
+
+  puts "Stack #{stack_exists ? 'exists' : 'does not exist'}, using #{operation}_stack"
+
+  # Prepare parameters for CloudFormation
+  cf_parameters = parameters.map {|k, v| {parameter_key: k, parameter_value: v.to_s}}
+  cf_tags = tags.map {|k, v| {key: k, value: v.to_s}}
 
-  # Add tags only if there are any
-  unless tags.empty?
-    tag_string = tags.map {|k, v| "#{k}=#{v}"}.join(" ")
-    command_parts << "--tags #{tag_string}"
+  # Build the CloudFormation request
+  request_params = {
+    stack_name: stack_name,
+    template_url: template_url,
+    parameters: cf_parameters,
+    tags: cf_tags,
+    capabilities: capabilities,
+    role_arn: cloudformation_role_arn
+  }.compact # Remove optional parameters like `role_arn` if not provided.
```

Comment:
would `template_body` be simpler than using `template_url` for the `create_stack` call? The nice thing about `cloudformation deploy` is that is handles this for us. The extra effort in this PR to send the template to S3 smells like unnecessary complexity to me. Is it necessary or carrying other benefits I'm not seeing?